src/utils/mod.rs

```diff
@@ -1010,3 +1011,164 @@ pub fn create_unrestricted_signing_rsa_public(
+            17 => Ok(PcrSlot::Slot17),
+            18 => Ok(PcrSlot::Slot18),
+            19 => Ok(PcrSlot::Slot19),
+            20 => Ok(PcrSlot::Slot20),
+            21 => Ok(PcrSlot::Slot21),
+            22 => Ok(PcrSlot::Slot22),
+            23 => Ok(PcrSlot::Slot23),
+            _ => Err(Error::local_error(WrapperErrorKind::InvalidParam)),
+        }
+    }
+}
+
+/// TPML_PCR_SELECTION is a structure that can contain up to
+/// 16 TPMS_PCR_SELECTION
+///
+/// The TPMS_PCR_SELECTION is a variable size bitmask
+/// where the position of each bit is the selected index.
+///
+/// The minimum number of octets in a TPMS_PCR_SELECT.sizeOfSelect
+/// is not determined by the number of PCR implemented but by the
+/// number of PCR required by the platform-specific
+/// specification with which the TPM is compliant or by the implementer if
+/// not adhering to a platform-specific specification.
+///
+/// So to make things simple 3 octets are
+/// used which means that size of selections always will be
+/// 3.
+// For example, selecting PCRs 3 and 9 looks like:
+// size(3)  mask     mask     mask     mask
+// 00000011 00000000 00000000 00000001 00000100
+#[derive(Debug, Default, Clone)]
+pub struct PcrSelections {
+    items: HashMap<HashingAlgorithm, HashSet<PcrSlot>>,
+}
+
+impl From<PcrSelections> for TPML_PCR_SELECTION {
+    fn from(pcr_selections: PcrSelections) -> TPML_PCR_SELECTION {
+        let mut ret: TPML_PCR_SELECTION = Default::default();
+        for (hash_algorithm, pcr_slots) in &pcr_selections.items {
+            ret.pcrSelections[ret.count as usize].hash = hash_algorithm.clone().into();
+            ret.pcrSelections[ret.count as usize].sizeofSelect = 3;
+            for &pcr_slot in pcr_slots {
+                let index: usize = (pcr_slot as usize) / 8;
+                let value: u8 = 1 << ((pcr_slot as u8) % 8);
+                ret.pcrSelections[ret.count as usize].pcrSelect[index] |= value;
+            }
+            ret.count += 1;
+        }
+        ret
+    }
+}
+
+impl From<TPML_PCR_SELECTION> for PcrSelections {
+    fn from(tpml_pcr_selection: TPML_PCR_SELECTION) -> PcrSelections {
+        let mut ret: PcrSelections = Default::default();
+        for selection_index in 0..(tpml_pcr_selection.count as usize) {
+            let selection = &tpml_pcr_selection.pcrSelections[selection_index];
+            let mut pcr_slots: HashSet<PcrSlot> = HashSet::<PcrSlot>::new();
+            for slot_nr in 1..(3 * 8) {
+                let index = slot_nr / 8;
+                let mask: u8 = (slot_nr as u8) % 8;
+                let is_set = (selection.pcrSelect[index] & mask) == mask;
+                if is_set {
+                    let _ = pcr_slots.insert(PcrSlot::try_from(slot_nr as u32).unwrap());
+                }
+            }
+            let _ = ret.items.insert(
+                HashingAlgorithm::try_from(selection.hash).unwrap(),
+                pcr_slots,
+            );
+        }
+        ret
+    }
+}
```

Comment:
I strongly suggest modifying this into a `TryFrom` because there are a few fields inside that structure that could be wrong and a `From` is strictly for cases where there is no possible failure or error (like in the `From` above where we control all inputs and we know they're safe - although that one has another issue as I noted) - in [docs](https://doc.rust-lang.org/std/convert/trait.From.html): "**Note: This trait must not fail.**"

You need to check that `tpml_pcr_selection.count`, for example, is a valid value. 

You're also expected to not use `.unwrap()` in cases where it's not provable that the result is `Ok(...)` - like when doing `HashingAlgorithm::try_from(...)`, as there's no guarantee that a user doesn't call this on a "bad" `TPML_PCR_SELECTION`. But that's easier to deal with if you swap for `TryFrom`

Reply:
Yea, I agree. I have fixed that now.

Reply:
Your are absolutely right. It was me being sloppy. It has been fixed now.

Reply:
Hey! Sorry for the late reply!

I've had a look, the swap to `TryFrom` is good, but it still needs the calls to `unwrap` to be changed to `?` - and for the case where you unwrap a `BitFlags::try_from` you can use [`or`](https://doc.rust-lang.org/std/result/enum.Result.html#method.or). Also, `selection_index` can trigger an "index out of bounds" panic as `tpml_pcr_selection.count` can be larger than the length of `tpml_pcr_selection.pcrSelections`